# original https://patch-diff.githubusercontent.com/raw/pulumi/pulumi/pull/12821.patch
From 30ee7475fba3eaef9551c63f9bfb12cad94f320b Mon Sep 17 00:00:00 2001
From: Dan Lorenc <dlorenc@chainguard.dev>
Date: Thu, 4 May 2023 07:10:19 -0400
Subject: [PATCH 1/2] Fix race condition in go/generate function.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

We were building pulumi and hit an error in this function when running the makefile, but only in certain environments.
I read the code, and it looks like there's a race condition here where the gofmt command finishes
before the io.Copy starts, giving us this error:

⚠️  x86_64    | + make build
ℹ️  x86_64    | BUILD:
ℹ️  x86_64    | go generate ./pulumi/...
⚠️  x86_64    | panic: unexpected error running gofmt: read |0: file already closed

The go async function is unnecessary here, and the Go exec docs explain that StderrPipe should not be used
with Run: https://pkg.go.dev/os/exec#Cmd.StderrPipe

CombinedOutput is a simpler way to accomplish the same thing, and fixes the race condition.

Signed-off-by: Dan Lorenc <dlorenc@chainguard.dev>
---
 sdk/go/pulumi/generate/main.go | 15 +++------------
 1 file changed, 3 insertions(+), 12 deletions(-)

diff --git a/sdk/go/pulumi/generate/main.go b/sdk/go/pulumi/generate/main.go
index 70cc5ce4876..4c751e7dc58 100644
--- a/sdk/go/pulumi/generate/main.go
+++ b/sdk/go/pulumi/generate/main.go
@@ -17,7 +17,6 @@ package main
 
 import (
 	"fmt"
-	"io"
 	"log"
 	"os"
 	"os/exec"
@@ -350,19 +349,11 @@ func main() {
 		f.Close()
 
 		gofmt := exec.Command("gofmt", "-s", "-w", fullname)
-		stderr, err := gofmt.StderrPipe()
+		out, err := gofmt.CombinedOutput()
 		if err != nil {
-			log.Fatalf("failed to pipe stderr from gofmt: %v", err)
-		}
-		go func() {
-			_, err := io.Copy(os.Stderr, stderr)
-			if err != nil {
-				panic(fmt.Sprintf("unexpected error running gofmt: %v", err))
-			}
-		}()
-		if err := gofmt.Run(); err != nil {
-			log.Fatalf("failed to gofmt %v: %v", fullname, err)
+			log.Fatalf("failed to gofmt %v: %v\n%s", fullname, err, string(out))
 		}
+
 	}
 }
 

From 3827f311c92c4599d03b31b410716cff7cbd3b40 Mon Sep 17 00:00:00 2001
From: Dan Lorenc <dlorenc@chainguard.dev>
Date: Thu, 4 May 2023 07:14:16 -0400
Subject: [PATCH 2/2] Add changelog entry.

Signed-off-by: Dan Lorenc <dlorenc@chainguard.dev>
---
 ...30504--build--fixes-race-condition-in-building-go-sdk.yaml | 4 ++++
 1 file changed, 4 insertions(+)
 create mode 100644 changelog/pending/20230504--build--fixes-race-condition-in-building-go-sdk.yaml

diff --git a/changelog/pending/20230504--build--fixes-race-condition-in-building-go-sdk.yaml b/changelog/pending/20230504--build--fixes-race-condition-in-building-go-sdk.yaml
new file mode 100644
index 00000000000..ab0a2da74c8
--- /dev/null
+++ b/changelog/pending/20230504--build--fixes-race-condition-in-building-go-sdk.yaml
@@ -0,0 +1,4 @@
+changes:
+- type: fix
+  scope: build
+  description: Fixes race condition in building Go sdk.
